#Data Ingestion

In [1]:
import os

In [2]:
%pwd

'c:\\Users\\SUPRIYA MISHRA\\OneDrive\\Desktop\\ML-MovieRec\\Movie-Recommender'

In [3]:
import pandas as pd
import numpy as np
 
movies = pd.read_csv('C:/Users/SUPRIYA MISHRA/OneDrive/Desktop/ML-MovieRec/tmdb_5000_movies.csv') 
credits = pd.read_csv('C:/Users/SUPRIYA MISHRA/OneDrive/Desktop/ML-MovieRec/tmdb_5000_credits.csv') 


#movies = pd.read_csv('/content/tmdb_5000_movies.csv')
#credits = pd.read_csv('/content/tmdb_5000_credits.csv')

In [4]:
movies.head()

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   
3  [{"name": "Legendary Pictures", "id": 923}, {"...   
4        [{"name": "Walt Disney Pictures", "id": 2}]   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...   2015-10-26   880674609   
3  [{"iso_3166_1": "US", "name": "United States o...   2012-07-16  1084939099   
4  [{"iso_3166_1": "US", "name": "United States o...   2012-03-07   284139100   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   
2    148.0  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3    165.0           [{"iso_639_1": "en", "name": "English"}]  Released   
4    132.0           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500  
2                                   Spectre           6.3        4466  
3                     The Dark Knight Rises           7.6        9106  
4                               John Carter           6.1        2124

In [5]:
credits.head(1)['crew'].values

array(['[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cam

merging both the dataframes.

In [6]:
movies.shape

(4803, 20)

In [7]:
credits.shape

(4803, 4)

In [8]:
movies = movies.merge(credits, on ='title')

In [9]:
print(movies.shape)
movies.head(1)

(4809, 23)


budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   

  original_title                                           overview  \
0         Avatar  In the 22nd century, a paraplegic Marine is di...   

   popularity                               production_companies  ... runtime  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...  ...   162.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   

                       tagline   title vote_average vote_count movie_id  \
0  Enter the World of Pandora.  Avatar          7.2      11800    19995   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  

[1 rows x 23 columns]

In [10]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4809 entries, 0 to 4808
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4809 non-null   int64  
 1   genres                4809 non-null   object 
 2   homepage              1713 non-null   object 
 3   id                    4809 non-null   int64  
 4   keywords              4809 non-null   object 
 5   original_language     4809 non-null   object 
 6   original_title        4809 non-null   object 
 7   overview              4806 non-null   object 
 8   popularity            4809 non-null   float64
 9   production_companies  4809 non-null   object 
 10  production_countries  4809 non-null   object 
 11  release_date          4808 non-null   object 
 12  revenue               4809 non-null   int64  
 13  runtime               4807 non-null   float64
 14  spoken_languages      4809 non-null   object 
 15  status               

we will select columns we want to keep and remove others.

genres, cast, id, crew, keywords, title, overview

not keeping original_title and mix of languages will cause a problem.

not keeping original language as english is a huge majority.

In [11]:
movies = movies[['genres','cast','movie_id','crew','keywords','title','overview']]

#Data Preprocessing

we will build a new dataframe with only three columns: id, title, and tags.

we will combine the remaining columns to create tag. But before that we will pre-process the columns and change their format, and do other necessary pre-processing tasks.

In [12]:
#missing data check
movies.isnull().sum()

genres      0
cast        0
movie_id    0
crew        0
keywords    0
title       0
overview    3
dtype: int64

In [13]:
#since only three movies have the missing data, remove the movies
movies.dropna(inplace=True)

In [14]:
movies.isnull().sum()

genres      0
cast        0
movie_id    0
crew        0
keywords    0
title       0
overview    0
dtype: int64

In [15]:
#check for duplicated
movies.duplicated().sum()

0

In [16]:
movies.iloc[0].genres #first row genre column

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]

we need this in the format:

['Action','Adventure','Fantasy','Science Fiction']

In [17]:
#also the output is a string of lists, we will convert it into a list with ast
import ast
def convert(obj):
  List = []
  for i in ast.literal_eval(obj):
    List.append(i['name'])
  return List

In [18]:
movies['genres']= movies['genres'].apply(convert)

In [19]:
movies.head() #head converted

genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                                cast  movie_id  \
0  [{"cast_id": 242, "character": "Jake Sully", "...     19995   
1  [{"cast_id": 4, "character": "Captain Jack Spa...       285   
2  [{"cast_id": 1, "character": "James Bond", "cr...    206647   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...     49026   
4  [{"cast_id": 5, "character": "John Carter", "c...     49529   

                                                crew  \
0  [{"credit_id": "52fe48009251416c750aca23", "de...   
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...   
2  [{"credit_id": "54805967c3a36829b5002c41", "de...   
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...   
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...   
4  [{"id": 818, "name": "based on novel"}, {"id":...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  
0  In the 22nd century, a paraplegic Marine is di...  
1  Captain Barbossa, long believed to be dead, ha...  
2  A cryptic message from Bond’s past sends him o...  
3  Following the death of District Attorney Harve...  
4  John Carter is a war-weary, former military ca...

In [20]:
movies['keywords']= movies['keywords'].apply(convert)

In [21]:
movies.head()

genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                                cast  movie_id  \
0  [{"cast_id": 242, "character": "Jake Sully", "...     19995   
1  [{"cast_id": 4, "character": "Captain Jack Spa...       285   
2  [{"cast_id": 1, "character": "James Bond", "cr...    206647   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...     49026   
4  [{"cast_id": 5, "character": "John Carter", "c...     49529   

                                                crew  \
0  [{"credit_id": "52fe48009251416c750aca23", "de...   
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...   
2  [{"credit_id": "54805967c3a36829b5002c41", "de...   
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...   
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  
0  In the 22nd century, a paraplegic Marine is di...  
1  Captain Barbossa, long believed to be dead, ha...  
2  A cryptic message from Bond’s past sends him o...  
3  Following the death of District Attorney Harve...  
4  John Carter is a war-weary, former military ca...

we want the top three dict for the cast, only the top three actors.

In [22]:
def convertCast(obj):
  List = []
  counter = 0
  for i in ast.literal_eval(obj):
    if counter < 3:
      List.append(i['name'])
      counter+=1
  return List

In [23]:
#movies['cast'] = movies['cast'].apply(lambda x:x[0:3])

In [24]:
movies['cast']= movies['cast'].apply(convertCast)

In [25]:
movies.head(1)

genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   

                                               cast  movie_id  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]     19995   

                                                crew  \
0  [{"credit_id": "52fe48009251416c750aca23", "de...   

                                            keywords   title  \
0  [culture clash, future, space war, space colon...  Avatar   

                                            overview  
0  In the 22nd century, a paraplegic Marine is di...

In [26]:
def fetch_director(obj):
  List = []
  counter = 0
  for i in ast.literal_eval(obj):
    if i['job'] == 'Director':
      List.append(i['name'])
  return List

In [27]:
movies['crew']= movies['crew'].apply(fetch_director)

In [28]:
movies.head()

genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                               cast  movie_id  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]     19995   
1     [Johnny Depp, Orlando Bloom, Keira Knightley]       285   
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]    206647   
3      [Christian Bale, Michael Caine, Gary Oldman]     49026   
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]     49529   

                  crew                                           keywords  \
0      [James Cameron]  [culture clash, future, space war, space colon...   
1     [Gore Verbinski]  [ocean, drug abuse, exotic island, east india ...   
2         [Sam Mendes]  [spy, based on novel, secret agent, sequel, mi...   
3  [Christopher Nolan]  [dc comics, crime fighter, terrorist, secret i...   
4     [Andrew Stanton]  [based on novel, mars, medallion, space travel...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  
0  In the 22nd century, a paraplegic Marine is di...  
1  Captain Barbossa, long believed to be dead, ha...  
2  A cryptic message from Bond’s past sends him o...  
3  Following the death of District Attorney Harve...  
4  John Carter is a war-weary, former military ca...

now we have to convert overview from a string to a list so that we can merge the other columns with it.

In [29]:
movies['overview']= movies['overview'].apply(lambda x:x.split())

In [30]:
movies.head()

genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                               cast  movie_id  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]     19995   
1     [Johnny Depp, Orlando Bloom, Keira Knightley]       285   
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]    206647   
3      [Christian Bale, Michael Caine, Gary Oldman]     49026   
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]     49529   

                  crew                                           keywords  \
0      [James Cameron]  [culture clash, future, space war, space colon...   
1     [Gore Verbinski]  [ocean, drug abuse, exotic island, east india ...   
2         [Sam Mendes]  [spy, based on novel, secret agent, sequel, mi...   
3  [Christopher Nolan]  [dc comics, crime fighter, terrorist, secret i...   
4     [Andrew Stanton]  [based on novel, mars, medallion, space travel...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...  
3  [Following, the, death, of, District, Attorney...  
4  [John, Carter, is, a, war-weary,, former, mili...

transformation to remove spaces.

In [31]:
movies['genres']= movies['genres'].apply(lambda x:[i.replace(" ", "") for i in x])
movies['keywords']= movies['keywords'].apply(lambda x:[i.replace(" ", "") for i in x])
movies['cast']= movies['cast'].apply(lambda x:[i.replace(" ", "") for i in x])
movies['crew']= movies['crew'].apply(lambda x:[i.replace(" ", "") for i in x])

In [32]:
movies.head()

genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   
1                  [Adventure, Fantasy, Action]   
2                    [Action, Adventure, Crime]   
3              [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, ScienceFiction]   

                                            cast  movie_id  \
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]     19995   
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]       285   
2      [DanielCraig, ChristophWaltz, LéaSeydoux]    206647   
3      [ChristianBale, MichaelCaine, GaryOldman]     49026   
4    [TaylorKitsch, LynnCollins, SamanthaMorton]     49529   

                 crew                                           keywords  \
0      [JamesCameron]  [cultureclash, future, spacewar, spacecolony, ...   
1     [GoreVerbinski]  [ocean, drugabuse, exoticisland, eastindiatrad...   
2         [SamMendes]  [spy, basedonnovel, secretagent, sequel, mi6, ...   
3  [ChristopherNolan]  [dccomics, crimefighter, terrorist, secretiden...   
4     [AndrewStanton]  [basedonnovel, mars, medallion, spacetravel, p...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...  
3  [Following, the, death, of, District, Attorney...  
4  [John, Carter, is, a, war-weary,, former, mili...

In [33]:
movies['tags'] = movies['overview']+ movies['genres'] + movies['keywords'] + movies['cast'] + movies['crew']

In [34]:
movies.head()

genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   
1                  [Adventure, Fantasy, Action]   
2                    [Action, Adventure, Crime]   
3              [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, ScienceFiction]   

                                            cast  movie_id  \
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]     19995   
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]       285   
2      [DanielCraig, ChristophWaltz, LéaSeydoux]    206647   
3      [ChristianBale, MichaelCaine, GaryOldman]     49026   
4    [TaylorKitsch, LynnCollins, SamanthaMorton]     49529   

                 crew                                           keywords  \
0      [JamesCameron]  [cultureclash, future, spacewar, spacecolony, ...   
1     [GoreVerbinski]  [ocean, drugabuse, exoticisland, eastindiatrad...   
2         [SamMendes]  [spy, basedonnovel, secretagent, sequel, mi6, ...   
3  [ChristopherNolan]  [dccomics, crimefighter, terrorist, secretiden...   
4     [AndrewStanton]  [basedonnovel, mars, medallion, spacetravel, p...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                                tags  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...  
3  [Following, the, death, of, District, Attorney...  
4  [John, Carter, is, a, war-weary,, former, mili...

In [35]:
new_df = movies[['movie_id', 'title', 'tags']]

In [36]:
new_df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...  
3  [Following, the, death, of, District, Attorney...  
4  [John, Carter, is, a, war-weary,, former, mili...

In [37]:
new_df['tags']= new_df['tags'].apply(lambda x:" ".join(x)) #join each element of tag with a space

C:\Users\SUPRIYA MISHRA\AppData\Local\Temp\ipykernel_9492\1605400965.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags']= new_df['tags'].apply(lambda x:" ".join(x)) #join each element of tag with a space


In [38]:
new_df.iloc[0].tags #is this correctly implemented?

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. Action Adventure Fantasy ScienceFiction cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d SamWorthington ZoeSaldana SigourneyWeaver JamesCameron'

#Text Vectorization

concept : Notion

we will remove the stop words from consideration.

In [39]:
!pip install nltk

In [40]:
import nltk #stemming
# example : ['love','loves','loved'] from totals words = ['love','love','love'] == ROOT WORDS CHOSEN

In [41]:
from nltk.stem.porter import PorterStemmer
ps = PorterStemmer()

In [42]:
def stem(text):
  y =[]
  for i in text.split(): #string to list
    y.append(ps.stem(i)) #stem every word from 5000 words
  return " ".join(y) #convert back to string

In [43]:
new_df['tags']= new_df['tags'].apply(stem)

C:\Users\SUPRIYA MISHRA\AppData\Local\Temp\ipykernel_9492\3021978705.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags']= new_df['tags'].apply(stem)


In [44]:
!pip install scikit-learn

In [45]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features = 5000, stop_words = 'english')


fit_transform: learns the vocabulary from new_df['tags'] and then transforms the text data into a numerical feature matrix.

.toarray()   : converts the sparse matrix (which is the default format returned by fit_transform for efficiency) into a dense NumPy array.


In [46]:
vectors = cv.fit_transform(new_df['tags']).toarray()

In [47]:
vectors.shape

(4806, 5000)

In [48]:
# Display the feature names (vocabulary) as a list
print(list(cv.get_feature_names_out()))

['000', '007', '10', '100', '11', '12', '13', '14', '15', '16', '17', '17th', '18', '18th', '18thcenturi', '19', '1910', '1920', '1930', '1940', '1944', '1950', '1950s', '1960', '1960s', '1970', '1970s', '1971', '1974', '1976', '1980', '1985', '1990', '1999', '19th', '19thcenturi', '20', '200', '2003', '2009', '20th', '21st', '23', '24', '25', '30', '300', '3d', '40', '50', '500', '60', '70', '80', 'aaron', 'aaroneckhart', 'abandon', 'abduct', 'abigailbreslin', 'abil', 'abl', 'aboard', 'abov', 'abus', 'academ', 'academi', 'accept', 'access', 'accid', 'accident', 'acclaim', 'accompani', 'accomplish', 'account', 'accus', 'ace', 'achiev', 'acquaint', 'act', 'action', 'actionhero', 'activ', 'activist', 'activities', 'actor', 'actress', 'actual', 'ad', 'adam', 'adamsandl', 'adamshankman', 'adapt', 'add', 'addict', 'adjust', 'admir', 'admit', 'adolesc', 'adopt', 'ador', 'adrienbrodi', 'adult', 'adultanim', 'adulteri', 'adulthood', 'advanc', 'adventur', 'adventure', 'adventures', 'advertis', 

In [49]:
vectors #will be sparse as out of 5000 only few will be filled

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

calculating the angle between the vectors using cosine distance

In [50]:
from sklearn.metrics.pairwise import cosine_similarity # cs is b/w 0 to 1, 1 closest. cosine_distances does the opposite

In [51]:
similarity = cosine_similarity(vectors) #distance of everymovie with everymovie

In [52]:
cosine_similarity(vectors).shape

(4806, 4806)

In [53]:
similarity

array([[1.        , 0.08346223, 0.0860309 , ..., 0.04499213, 0.        ,
        0.        ],
       [0.08346223, 1.        , 0.06063391, ..., 0.02378257, 0.        ,
        0.02615329],
       [0.0860309 , 0.06063391, 1.        , ..., 0.02451452, 0.        ,
        0.        ],
       ...,
       [0.04499213, 0.02378257, 0.02451452, ..., 1.        , 0.03962144,
        0.04229549],
       [0.        , 0.        , 0.        , ..., 0.03962144, 1.        ,
        0.08714204],
       [0.        , 0.02615329, 0.        , ..., 0.04229549, 0.08714204,
        1.        ]])

Main Function to recommend most similar 5 movies.

In [54]:
new_df[new_df['title']== 'Batman Begins']

movie_id          title  \
119       272  Batman Begins   

                                                  tags  
119  driven by tragedy, billionair bruce wayn dedic...

In [55]:
new_df[new_df['title']== 'Batman Begins'].index[0]

119

In [56]:
#enumerate function will add index along with the movies while we sort them on the basis of the second value of the tuple.
sorted(list(enumerate(similarity[0])), reverse = True, key= lambda x:x[1]) #for the 0th index movie

[(0, 1.0000000000000002),
 (1216, 0.28676966733820225),
 (2409, 0.26901379342448517),
 (3730, 0.2605130246476754),
 (507, 0.255608593705383),
 (539, 0.25038669783359574),
 (582, 0.24511108480187255),
 (1204, 0.24455799402225925),
 (1194, 0.23179316248638276),
 (778, 0.23174488732966073),
 (4048, 0.2278389747471728),
 (1920, 0.2252817784447915),
 (61, 0.22269966704152225),
 (2786, 0.21853668936906193),
 (172, 0.21239769762143662),
 (972, 0.2108663315950723),
 (322, 0.2105263157894737),
 (2333, 0.20443988269091456),
 (3608, 0.20437977982832192),
 (260, 0.20395079136182276),
 (151, 0.2029530274475215),
 (4192, 0.2029530274475215),
 (1444, 0.20277677641345318),
 (74, 0.2024645717996314),
 (1089, 0.2020475485519274),
 (3675, 0.1979082783981174),
 (973, 0.19767387315371682),
 (577, 0.1976738731537168),
 (47, 0.19672236884115843),
 (2971, 0.1925214071641298),
 (942, 0.19134594929397597),
 (495, 0.19088542889273336),
 (1201, 0.19088542889273336),
 (305, 0.19007487139298027),
 (4, 0.18929940971

In [57]:
sorted(list(enumerate(similarity[0])), reverse = True, key= lambda x:x[1])[1:6]

[(1216, 0.28676966733820225),
 (2409, 0.26901379342448517),
 (3730, 0.2605130246476754),
 (507, 0.255608593705383),
 (539, 0.25038669783359574)]

In [58]:
def recommend(movie):
  index = new_df[new_df['title'] == movie].index[0]
  distances = sorted(list(enumerate(similarity[index])),reverse=True,key = lambda x: x[1])
  for i in distances[1:6]:
    print(new_df.iloc[i[0]].title)

In [59]:
recommend('Batman Begins')

The Dark Knight
Batman
Batman
The Dark Knight Rises
10th & Wolf


#Import

In [60]:
import pickle


pickle.dump(new_df.to_dict(), open('C:/Users/SUPRIYA MISHRA/OneDrive/Desktop/ML-MovieRec/movies_dict1.pkl', 'wb'))
pickle.dump(similarity, open('C:/Users/SUPRIYA MISHRA/OneDrive/Desktop/ML-MovieRec/similarity1.pkl', 'wb'))


In [61]:
#pickle.dump(new_df.to_dict,open('movies_dict.pkl', 'wb'))

In [62]:
#pickle.dump(similarity,open('similarity.pkl','wb'))